### PR TITLE
102 performant send

### LINKED ISSUE #102: performant send
actually the major client (java/C#/go) use a sort of internal queue where all the message sent are queued and every x millis a send batch is performed (to meet performance).

this queue (for example 100 message) needs to meet some requirements (as ex the requestMaxFrameSize dimension) see the java or go client to get some inspiration

### PR DESCRIPTION
**Pull request by icappello**
closes #102 

Based on PR #119

### CODE REVIEW

File: src/requests/abstract_request.ts [hunk 1]
```diff
@@ -25,54 +26,88 @@ export class BufferDataWriter implements DataWriter {
...
     this._offset = this.buffer.writeInt16BE(data.length, this._offset)
     this.writeData(data)
   }
 
   toBuffer(): Buffer {
     return this.buffer.slice(0, this._offset)
   }
+
+  private growIfNeeded(additionalBytes: number) {
+    const maxRatio = 0.9
```
Comment (tarzacodes) [inline src/requests/abstract_request.ts:88]: magari questo maxRatio lo possiamo rendere configurabile, per giocarci un poco almeno nei test?

File: src/requests/abstract_request.ts [hunk 2]
```diff
@@ -25,54 +26,88 @@ export class BufferDataWriter implements DataWriter {
...
+    if ((this._offset + additionalBytes) / this.buffer.length > maxRatio) {
+      this.growBuffer(additionalBytes)
+    }
+  }
+
+  private growBuffer(requiredBytes: number) {
+    const newSize = Math.min(this.buffer.length * 2 + requiredBytes, this.writeBufferMaxSize)
```
[the remark this answers is not in the document]
Comment (icappello) [reply src/requests/abstract_request.ts:95]: yes, it's a basic heuristic to balance memory usage vs new buffer allocations (especially when dealing with big payloads)

**Review by albertobarrila — DISMISSED** [review]
LGTM

File: performance_test/perf_test_producer.ts [hunk 3]
```diff
@@ -27,36 +27,39 @@ export class PerfTestProducer {
...
+    return true
   }
 
   private displayMetrics(stop: boolean = false) {
     const metrics = { ...this.metrics.getMetrics(), total: this.ctr }
-    this.logger.info(`${new Date().toISOString()} - ${inspect(metrics)}`)
+    this.logger.info(`${inspect(metrics)}`)
     if (stop && this.displayTimer) {
-      clearInterval(this.displayTimer)
+      clearInterval(this.displayTimer[Symbol.toPrimitive]())
```
Comment (tarzacodes) [inline performance_test/perf_test_producer.ts:49]: Never encountered this, how does this line work? Just wondering

Comment (icappello) [reply performance_test/perf_test_producer.ts]: it's a way to get the id of the timer. Fixing the type definition to avoid this

**Review by albertobarrila — APPROVED** [review]
:+1:

**Pull request merged**